smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaRecordStream.java

```diff
@@ -82,14 +90,18 @@ public KafkaRecordStreamSubscription(
             this.client = client;
             this.pauseResumeEnabled = config.getPauseIfNoRequests();
             this.downstream = subscriber;
+            this.size = new AtomicLong(0);
             this.queue = new ConcurrentLinkedDeque<>();
             this.retries = config.getRetryAttempts() == -1 ? Long.MAX_VALUE : config.getRetryAttempts();
             this.pollUni = client.poll()
                     .onItem().transform(cr -> {
                         if (cr.isEmpty()) {
                             return null;
                         }
-                        cr.forEach(queue::offer);
+                        for (ConsumerRecord<K, V> rec : cr) {
+                            queue.offer(rec);
+                        }
+                        size.addAndGet(cr.count());
```

Comment:
I don't truly understand the code, so please treat everything I write here as nothing more than a comment... but this seems rather racy. You could for example easily end up with a negative `size`, which would be unexpected. But the code doesn't seem to care. And, I mean, it was racy before, so it's probably fine :-)

If `ConsumerRecords` was a `Collection` (it's `Iterable` :-/), we could `queue.addAll(cr)`, which has a single linearization point and returns shortly after it, and the race window would be much smaller.

(Interestingly, `ConcurrentLinkedDeque.addAll` could easily take an `Iterable`, because it doesn't use anything else from the ` Collection` API. But, of course, Java doesn't support contravariant parameters.)

Alternatively, we could increment after each `queue.offer`, which again makes the race window a lot smaller (at the cost of the `AtomicLong` overhead).

Reply:
Yes, there is a race. Unfortunately, there is no good solution except re-implementing a Queue. I will try this.

Reply:
Just pushed a different approach using a specialized queue.